pinot-plugins/pinot-stream-ingestion/pinot-kinesis/src/main/java/org/apache/pinot/plugin/stream/kinesis/KinesisStreamMetadataProvider.java

```diff
@@ -0,0 +1,177 @@
+   */
+  @Override
+  public List<PartitionGroupMetadata> computePartitionGroupMetadata(String clientId, StreamConfig streamConfig,
+      List<PartitionGroupConsumptionStatus> partitionGroupConsumptionStatuses, int timeoutMillis)
+      throws IOException, TimeoutException {
+
+    List<PartitionGroupMetadata> newPartitionGroupMetadatas = new ArrayList<>();
+
+    Map<String, Shard> shardIdToShardMap = _kinesisConnectionHandler.getShards().stream()
+        .collect(Collectors.toMap(Shard::shardId, s -> s, (s1, s2) -> s1));
+    Set<String> shardsInCurrent = new HashSet<>();
+    Set<String> shardsEnded = new HashSet<>();
+
+    // TODO: Once we start supporting multiple shards in a PartitionGroup,
+    //  we need to iterate over all shards to check if any of them have reached end
+
+    // Process existing shards. Add them to new list if still consuming from them
+    for (PartitionGroupConsumptionStatus currentPartitionGroupConsumptionStatus : partitionGroupConsumptionStatuses) {
+      KinesisPartitionGroupOffset kinesisStartCheckpoint =
+          (KinesisPartitionGroupOffset) currentPartitionGroupConsumptionStatus.getStartOffset();
+      String shardId = kinesisStartCheckpoint.getShardToStartSequenceMap().keySet().iterator().next();
+      Shard shard = shardIdToShardMap.get(shardId);
+      shardsInCurrent.add(shardId);
+
+      StreamPartitionMsgOffset newStartOffset;
+      StreamPartitionMsgOffset currentEndOffset = currentPartitionGroupConsumptionStatus.getEndOffset();
+      if (currentEndOffset != null) { // Segment DONE (committing/committed)
+        String endingSequenceNumber = shard.sequenceNumberRange().endingSequenceNumber();
+        if (endingSequenceNumber != null) { // Shard has ended, check if we're also done consuming it
+          if (consumedEndOfShard(currentEndOffset, currentPartitionGroupConsumptionStatus)) {
+            shardsEnded.add(shardId);
+            continue; // Shard ended and we're done consuming it. Skip
+          }
+        }
+        newStartOffset = currentEndOffset;
+      } else { // Segment IN_PROGRESS
+        newStartOffset = currentPartitionGroupConsumptionStatus.getStartOffset();
+      }
+      newPartitionGroupMetadatas.add(
+          new PartitionGroupMetadata(currentPartitionGroupConsumptionStatus.getPartitionGroupId(), newStartOffset));
+    }
+
+    // Add new shards. Parent should be null (new table case, very first shards)
+    // OR it should be flagged as reached EOL and completely consumed.
+    for (Map.Entry<String, Shard> entry : shardIdToShardMap.entrySet()) {
+      String newShardId = entry.getKey();
+      if (shardsInCurrent.contains(newShardId)) {
+        continue;
+      }
+      StreamPartitionMsgOffset newStartOffset;
+      Shard newShard = entry.getValue();
+      String parentShardId = newShard.parentShardId();
+
+      if (parentShardId == null || shardsEnded.contains(parentShardId)) {
```

Comment:
Can there be an edge case for this check? 

(e.g. Segments with the parent shard id got deleted due to retention before we reach this code path)

Reply:
fixed in the new shards case.
Thought of another edge case (shard expired due to retention but the consumer was not done consuming it). That's a more tricky case, and i've added a todo for now, because i dont know if practically we'll reach it.